storage/rocksdb/clone/donor.cc

```diff
@@ -0,0 +1,1406 @@
+    return clone::return_error(ER_CLONE_PROTOCOL, "Invalid clone locator");
+
+  assert(donor_instance->get_state() ==
+         donor::donor_state::FINAL_CHECKPOINT_WITH_LOGS);
+
+  const auto err = donor_instance->handle_any_error(thd);
+  if (err != 0) return err;
+
+  cbk->set_hton(hton);
+  return donor_instance->copy(thd, task_id, *cbk);
+}
+
+// Acknowledge clone application errors that are received from the client. In
+// general, this callback is used to acknowledge the client advancing to the
+// next clone stage, but this capability is unused in MyRocks clone.
+int rocksdb_clone_ack(handlerton *, THD *thd, const uchar *loc, uint loc_len,
+                      uint task_id, int in_err, Ha_clone_cbk *) {
+  const auto l = myrocks::clone::locator::deserialize(loc, loc_len);
+  if (!l.is_valid() || l.has_applier_status())
+    return clone::return_error(ER_CLONE_PROTOCOL, "Invalid clone locator");
+  auto *const donor_instance =
+      lookup_donor_and_task_id(l, loc, loc_len, task_id);
+  if (donor_instance == nullptr)
+    return clone::return_error(ER_CLONE_PROTOCOL, "Invalid clone locator");
+
+  // ACK is only used to report errors for MyRocks clone
+  if (in_err == 0) {
+    return clone::return_error(ER_CLONE_PROTOCOL,
+                               "Invalid ACK received from client");
+  }
+
+  if (thd_killed(thd)) {
+    my_error(ER_QUERY_INTERRUPTED, MYF(0));
+    in_err = ER_QUERY_INTERRUPTED;
+  }
+
+  donor_instance->save_error(in_err);
+
+  return in_err;
+}
+
+// Finish the donor-side clone session for this storage engine.
+int rocksdb_clone_end(handlerton *, THD *thd, const uchar *loc, uint loc_len,
+                      uint task_id, int in_err) {
+  const auto end_locator = myrocks::clone::locator::deserialize(loc, loc_len);
+  if (!end_locator.is_valid() || end_locator.has_applier_status())
+    return clone::return_error(ER_CLONE_PROTOCOL, "Invalid clone locator");
+
+  auto *const donor_instance =
+      lookup_donor_and_task_id(end_locator, loc, loc_len, task_id);
+  if (donor_instance == nullptr)
+    return clone::return_error(ER_CLONE_PROTOCOL, "Invalid clone locator");
+
+  const auto is_master_task_id = clone::session::is_master_task_id(task_id);
+  if (is_master_task_id) manager::instance().reject_new_tasks(*donor_instance);
+
+  const auto err = donor_instance->get_error();
+
+  if (thd_killed(thd)) {
+    my_error(ER_QUERY_INTERRUPTED, MYF(0));
+    in_err = ER_QUERY_INTERRUPTED;
+  } else if (err == ER_QUERY_INTERRUPTED) {
+    in_err = err;
+  }
+  donor_instance->save_error(in_err);
+
+  const auto clone_successful = in_err == 0;
+
+  if (clone_successful) {
+    sql_print_information("Clone end: task ID: %u, success", task_id);
+  } else {
+    sql_print_information("Clone end: task ID: %u, failed with code %d: %s",
+                          task_id, in_err, clone::get_da_error_message(*thd));
+  }
+
+  if (is_master_task_id) {
+    donor_instance->wait_until_one_task();
```

Comment:
What happens during clone cleanup? Does each client task thread issue COM_EXIT to its corresponding donor task thread? It looks like `hton_clone_end` loops through all tasks too when processing COM_EXIT, which seems odd if each thread is also issuing COM_EXIT.

Does blocking here for master_task_id and waiting for all tasks to terminate potentially introduce a deadlock if `hton_clone_end` is looping through all tasks in a single thread starting with the task_id 0?

Also, while upstream code does use master_task_id, preference would be to change the terminology to coordinator or primary or something similar.

Reply:
> What happens during clone cleanup? Does each client task thread issue COM_EXIT to its corresponding donor task thread?

Yes.

>  It looks like `hton_clone_end` loops through all tasks too when processing COM_EXIT, which seems odd if each thread is also issuing COM_EXIT.
> 
> Does blocking here for master_task_id and waiting for all tasks to terminate potentially introduce a deadlock if `hton_clone_end` is looping through all tasks in a single thread starting with the task_id 0?

`hton_clone_end` does not loop over the global tasks--its `task_vec` arg is one thread's tasks across storage engines.

> Also, while upstream code does use master_task_id, preference would be to change the terminology to coordinator or primary or something similar.

I am worried about the confusion from naming the same thing differently, but I used "main"/"worker" in the Wiki already. I'll go with that in the code too if that's OK